### PR TITLE
Allow tuned lockdown confidentiality

### PR DESCRIPTION
**Pull request by zpytela**

### CODE REVIEW

**Comment by WOnder93** [comment]
Was this tested in permissive mode? The confidentiality denial seems to be from `perf_event_open(2)`, so I'd expected some `perf_event` AVCs as well.

**Comment by ghost** [comment]
Ondrej Mosnáček ***@***.***> writes:

> Was this tested in permissive mode? The confidentiality denial seems to be from perf_event_open(2), so
> I'd expected some perf_event AVCs as well.

I am surprised anything showed up as its a unconfined domain. I guess
unconfined domains dont have unconfined access to lockdown

>
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly, view it on GitHub, or unsubscribe.
>

-- 
gpg --locate-keys ***@***.***
Key fingerprint = FCD2 3660 5D6B 9D27 7FC6  E0FF DA7E 521F 10F6 4098
https://sks-keyservers.net/pks/lookup?op=get&search=0xDA7E521F10F64098
Dominick Grift

**Comment by WOnder93** [comment]
Ah, right, that explains it... So that commit should either be dropped in favor of #720 or should add the appropriate `kernel_*_perf_event()` interface call instead (or both).

**Comment by zpytela** [comment]
I haven't noticed tuned_t is unconfined_domain_type which explains the single avc, so this is just for the record, but I want to share the results as I finally managed to reproduce it:
```
type=PROCTITLE msg=audit(05/07/2021 04:52:04.295:189) : proctitle=/usr/bin/python3 -Es /usr/sbin/tuned -l -P
type=SYSCALL msg=audit(05/07/2021 04:52:04.295:189) : arch=x86_64 syscall=perf_event_open success=yes exit=10 a0=0x7f745801b480 a1=0xffffffff a2=0x0 a3=0xffffffff items=0 ppid=1 pid=593 auid=unset uid=root gid=root euid=root suid=root fsuid=root egid=root sgid=root fsgid=root tty=(none) ses=unset comm=tuned exe=/usr/bin/python3.9 subj=system_u:system_r:tuned_t:s0 key=(null)
type=AVC msg=audit(05/07/2021 04:52:04.295:189) : avc:  denied  { confidentiality } for  pid=593 comm=tuned lockdown_reason="unsafe use of perf" scontext=system_u:system_r:tuned_t:s0 tcontext=system_u:system_r:tuned_t:s0 tclass=lockdown permissive=1
```
This is the only AVC which appears, but after removing the unconfined module there are others:
```
type=PROCTITLE msg=audit(05/07/2021 04:56:43.940:643) : proctitle=/usr/bin/python3 -Es /usr/sbin/tuned -l -P
type=SYSCALL msg=audit(05/07/2021 04:56:43.940:643) : arch=x86_64 syscall=perf_event_open success=no exit=EACCES(Permission denied) a0=0x7fb4c0019ed0 a1=0xffffffff a2=0x0 a3=0xffffffff items=0 ppid=1 pid=3391 auid=unset uid=root gid=root euid=root suid=root fsuid=root egid=root sgid=root fsgid=root tty=(none) ses=unset comm=tuned exe=/usr/bin/python3.9 subj=system_u:system_r:tuned_t:s0 key=(null)
type=AVC msg=audit(05/07/2021 04:56:43.940:643) : avc:  denied  { open } for  pid=3391 comm=tuned scontext=system_u:system_r:tuned_t:s0 tcontext=system_u:system_r:tuned_t:s0 tclass=perf_event permissive=0
----
type=PROCTITLE msg=audit(05/07/2021 04:56:43.953:648) : proctitle=/usr/bin/python3 -Es /usr/sbin/tuned -l -P
type=PATH msg=audit(05/07/2021 04:56:43.953:648) : item=1 name=/etc/tuned/bootcmdline inode=41943371 dev=fd:01 mode=file,644 ouid=root ogid=root rdev=00:00 obj=system_u:object_r:tuned_etc_t:s0 nametype=NORMAL cap_fp=none cap_fi=none cap_fe=0 cap_fver=0 cap_frootid=0
type=PATH msg=audit(05/07/2021 04:56:43.953:648) : item=0 name=/etc/tuned/ inode=41943369 dev=fd:01 mode=dir,755 ouid=root ogid=root rdev=00:00 obj=system_u:object_r:tuned_etc_t:s0 nametype=PARENT cap_fp=none cap_fi=none cap_fe=0 cap_fver=0 cap_frootid=0
type=CWD msg=audit(05/07/2021 04:56:43.953:648) : cwd=/
type=SYSCALL msg=audit(05/07/2021 04:56:43.953:648) : arch=x86_64 syscall=openat success=no exit=EACCES(Permission denied) a0=0xffffff9c a1=0x7fb4c5c28290 a2=O_WRONLY|O_CREAT|O_TRUNC|O_CLOEXEC a3=0x1b6 items=2 ppid=1 pid=3391 auid=unset uid=root gid=root euid=root suid=root fsuid=root egid=root sgid=root fsgid=root tty=(none) ses=unset comm=tuned exe=/usr/bin/python3.9 subj=system_u:system_r:tuned_t:s0 key=(null)
type=AVC msg=audit(05/07/2021 04:56:43.953:648) : avc:  denied  { write } for  pid=3391 comm=tuned name=bootcmdline dev="vda1" ino=41943371 scontext=system_u:system_r:tuned_t:s0 tcontext=system_u:object_r:tuned_etc_t:s0 tclass=file permissive=0
```
The complete list of denials to address is about 80 entries.

File: policy/modules/contrib/tuned.te [hunk 1]
```diff
@@ -41,7 +41,7 @@ allow tuned_t self:netlink_socket create_socket_perms;
 allow tuned_t self:udp_socket create_socket_perms;
 allow tuned_t self:socket create_socket_perms;
 
-read_files_pattern(tuned_t, tuned_etc_t, tuned_etc_t)
+rw_files_pattern(tuned_t, tuned_etc_t, tuned_etc_t)
```
Comment (WOnder93) [inline policy/modules/contrib/tuned.te]: I can see `tuned_rw_etc_t` below, which seems to be intended for rw files. So I believe we should rather change the label for the file(s) that tuned needs to write to.

File: policy/modules/system/userdomain.te [hunk 2]
```diff
@@ -206,6 +206,10 @@ optional_policy(`
 	telepathy_filetrans_home_content(userdomain)
 ')
 
+optional_policy(`
+	tuned_dbus_chat(userdomain)
```
Comment (WOnder93) [inline policy/modules/system/userdomain.te]: I wonder if this should really be allowed to the whole `userdomain`... Communicating with tuned sounds like something only sysadm_t (maybe also staff_t?) should be allowed to do.

Comment (zpytela) [reply policy/modules/system/userdomain.te]: It is only sysadm_t, staff_u only in the sysadm_r role. So the tests needs updating, too, apparently not everything which is tested is reasonable.

File: policy/modules/contrib/tuned.te (hunk 1 above)
Comment (zpytela) [reply policy/modules/contrib/tuned.te]: Looks safe, the only interface making tuned_rw_etc_t accessible is tuned_admin. So I found all 4 instances of changed files, and while there, fixed the incorrect existing transition.

**Review by WOnder93 — APPROVED** [review]
Looks good now :+1:

**Comment by zpytela** [comment]
The perf_event write permissions are not sufficient, also perfmon capability is required which is covered by the `kernel_manage_perf_event` interface usage. Additionally, I added rw_files pattern for tuned_etc_t and decided not to cover all the remaining permissions.
```
----
type=PROCTITLE msg=audit(05/07/2021 05:03:46.333:939) : proctitle=/usr/bin/python3 -Es /usr/sbin/tuned -l -P
type=SYSCALL msg=audit(05/07/2021 05:03:46.333:939) : arch=x86_64 syscall=perf_event_open success=yes exit=8 a0=0x7fcbf401db80 a1=0xffffffff a2=0x0 a3=0xffffffff items=0 ppid=1 pid=3481 auid=unset uid=root gid=root euid=root suid=root fsuid=root egid=root sgid=root fsgid=root tty=(none) ses=unset comm=tuned exe=/usr/bin/python3.9 subj=system_u:system_r:tuned_t:s0 key=(null)
type=AVC msg=audit(05/07/2021 05:03:46.333:939) : avc:  denied  { cpu } for  pid=3481 comm=tuned scontext=system_u:system_r:tuned_t:s0 tcontext=system_u:system_r:tuned_t:s0 tclass=perf_event permissive=1
type=AVC msg=audit(05/07/2021 05:03:46.333:939) : avc:  denied  { kernel } for  pid=3481 comm=tuned scontext=system_u:system_r:tuned_t:s0 tcontext=system_u:system_r:tuned_t:s0 tclass=perf_event permissive=1
type=AVC msg=audit(05/07/2021 05:03:46.333:939) : avc:  denied  { perfmon } for  pid=3481 comm=tuned capability=unknown-capability(38)  scontext=system_u:system_r:tuned_t:s0 tcontext=system_u:system_r:tuned_t:s0 tclass=capability2 permissive=1
type=AVC msg=audit(05/07/2021 05:03:46.333:939) : avc:  denied  { open } for  pid=3481 comm=tuned scontext=system_u:system_r:tuned_t:s0 tcontext=system_u:system_r:tuned_t:s0 tclass=perf_event permissive=1
```

**Pull request merged**